### ISSUE TITLE
Failed to backmerge branches.: xJ: Command failed with ENOENT: git fetch

### ISSUE THREAD

**Issue by syphernl** [kilianpaquier/semantic-release-backmerge#17]
We are running in to a rather odd issue with this plugin. In one particular project the backmerges always fail with:

```
Failed to backmerge branches.

xJ: Command failed with ENOENT: git fetch 'https://gitlab-ci-token:[secure]@gitlab.com/REDACTED' spawnSync git ENOENT
```

We have `GITLAB_TOKEN` set as Masked+Hidden variable. The token itself is fine, since `semantic-release/gitlab` can create releases + tags just fine using it.
Git is installed in the container as well:

```bash
~ # git --version
git version 2.45.2
```

Relevant part of `.releaserc.js`:

```js
    [
      "@kilianpaquier/semantic-release-backmerge",
      {
        targets: [
          { from: "production", to: "develop" },
          { from: "acceptance", to: "develop" }
        ],
        title: "ci(release): fix mergeback from ${ from } into ${ to } [skip ci]",
      }
    ],
```

Logs:

```bash
[8:16:48 AM] [semantic-release] › ℹ  Start step "success" of plugin "@kilianpaquier/semantic-release-backmerge"
[8:16:48 AM] [semantic-release] [@kilianpaquier/semantic-release-backmerge] › ℹ  Current branch 'acceptance' matches following configured backmerge targets: '[{"from":"acceptance","to":"develop"}]'. Performing backmerge.
[8:16:48 AM] [semantic-release] › ✘  Failed step "success" of plugin "@kilianpaquier/semantic-release-backmerge"
[8:16:48 AM] [semantic-release] › ℹ  Start step "fail" of plugin "@semantic-release/gitlab"
[8:16:49 AM] [semantic-release] [@semantic-release/gitlab] › ℹ  Commented on issue REDACTED: https://gitlab.com/REDACTED/-/issues/X.
[8:16:49 AM] [semantic-release] › ✔  Completed step "fail" of plugin "@semantic-release/gitlab"
[8:16:49 AM] [semantic-release] › ℹ  Start step "fail" of plugin "semantic-release-slack-bot"
[8:16:49 AM] [semantic-release] [semantic-release-slack-bot] › ℹ  Notifying on fail skipped
[8:16:49 AM] [semantic-release] › ✔  Completed step "fail" of plugin "semantic-release-slack-bot"
[8:16:49 AM] [semantic-release] › ✘  EBACKMERGE Failed to backmerge branches.
Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'marked-terminal' imported from /usr/local/lib/node_modules/semantic-release/index.js
    at packageResolve (node:internal/modules/esm/resolve:854:9)
    at moduleResolve (node:internal/modules/esm/resolve:927:18)
    at defaultResolve (node:internal/modules/esm/resolve:1169:11)
    at ModuleLoader.defaultResolve (node:internal/modules/esm/loader:542:12)
    at ModuleLoader.resolve (node:internal/modules/esm/loader:510:25)
    at ModuleLoader.getModuleJob (node:internal/modules/esm/loader:239:38)
    at ModuleLoader.import (node:internal/modules/esm/loader:472:34)
    at defaultImportModuleDynamicallyForModule (node:internal/modules/esm/utils:214:31)
    at importModuleDynamicallyCallback (node:internal/modules/esm/utils:253:12)
    at terminalOutput (file:///usr/local/lib/node_modules/semantic-release/index.js:29:43) {
  code: 'ERR_MODULE_NOT_FOUND'
}
```

I have tried to reproduce the issue in a separate repository, but for the backmerges seem to work fine there. They both use the same config in `.releaserc.js`. Is there a way to get more logging out of this plugin to see what it is actually doing and why it is failing?

**Comment by kilianpaquier** [comment]
Hello, thank you for opening this issue !

You can enable semantic-release in debug mode (`--debug`) but I don't think I added debug logs.

Some questions :
- What's the semantic-release version you're using ?
- The first provided error isn't in the logs at the end of your issue, are they related ?

**Comment by syphernl** [comment]
Hi @kilianpaquier, thanks for your quick reply :)

> * What's the semantic-release version you're using ?

We run version `24.1.2`.

> * The first provided error isn't in the logs at the end of your issue, are they related ?

The first error (and the title) is the error that showed up in the "Release is failing" issue created by Semantic-Release.

> * What OS is being used for this job ?

We run this job in a container based on `Alpine 3.20.3`.

It is rather strange that it is failing with one project, but not with the other. 
I have doublechecked and compared GitLab settings, pipeline configs, access token settings etc and there aren't any differences as far as I can see. The ENOENT seems to indicate it cannot run `git` (because it does not exist) but it is part of the container and works fine in another project..

**Comment by kilianpaquier** [comment]
Thank you for this reply.

Could you try installing `marked-terminal` manually during the CI with `npm i -g marked-terminal` and rerun the thing ? 

You can run in `--dry-run` since it'll at least fetch and run a `git push --dry-run`. If the issue with `git` is still there it'll fail.

For the dry run to go through the whole process, you'll at least need to have a commit eligible to releasing 🙂.

**Comment by syphernl** [comment]
In the `before_script` I added the install command, but the error with `marked-terminal` is still the same. 

Could this ENOENT problem be permission related? 
If it is using the output from `git ls-remote --heads` it will get an URL which Gitlab CI used which includes the (short-lived and limited) [CI_JOB_TOKEN](https://docs.gitlab.co.jp/ee/ci/jobs/ci_job_token.html) that does not have write permissions, while the `GITLAB_TOKEN ` is expected to have that.
Still weird, since the error it claims to be failing on is `git fetch` which is allowed via the `CI_JOB_TOKEN`..

Running with env `DEBUG=semantic-release:*` didn't show any info for the plugin (which was expected as you indicated earlier) but I did see this:

```bash
2024-10-21T11:46:22.017Z semantic-release:get-git-auth-url Verifying ssh auth by attempting to push to  https://gitlab-ci-token:[secure]@gitlab.com/REDACTED.git
2024-10-21T11:46:22.510Z semantic-release:git ExecaError: Command failed with exit code 128: git push --dry-run --no-verify 'https://gitlab-ci-token:[secure]@gitlab.com/REDACTED.git' 'HEAD:acceptance'
remote: You are not allowed to push code to this project.
fatal: unable to access 'https://gitlab.com/REDACTED.git/': The requested URL returned error: 403
    at getFinalError (file:///usr/local/lib/node_modules/semantic-release/node_modules/execa/lib/return/final-error.js:6:9)
    at makeError (file:///usr/local/lib/node_modules/semantic-release/node_modules/execa/lib/return/result.js:108:16)
    at getAsyncResult (file:///usr/local/lib/node_modules/semantic-release/node_modules/execa/lib/methods/main-async.js:167:4)
    at handlePromise (file:///usr/local/lib/node_modules/semantic-release/node_modules/execa/lib/methods/main-async.js:150:17)
    at async verifyAuth (file:///usr/local/lib/node_modules/semantic-release/lib/git.js:206:5)
    at async default (file:///usr/local/lib/node_modules/semantic-release/lib/get-git-auth-url.js:93:5)
    at async run (file:///usr/local/lib/node_modules/semantic-release/index.js:67:27)
    at async Module.default (file:///usr/local/lib/node_modules/semantic-release/index.js:278:22)
    at async default (file:///usr/local/lib/node_modules/semantic-release/cli.js:55:5)
2024-10-21T11:46:22.511Z semantic-release:get-git-auth-url SSH key auth failed, falling back to https.
2024-10-21T11:46:30.600Z semantic-release:get-tags found tags for branch review: []
2024-10-21T11:46:31.097Z semantic-release:get-tags found tags for branch acceptance: [ <list_of_tags_here> ]
2024-10-21T11:46:31.490Z semantic-release:get-tags found tags for branch production: [ <list_of_tags_here> ]
[11:46:31 AM] [semantic-release] › ✔  Run automated release from branch acceptance on repository https://gitlab-ci-token:[secure]@gitlab.com/REDACTED.git
[11:46:32 AM] [semantic-release] › ✔  Allowed to push to the Git repository
```

**Comment by syphernl** [comment]
If I do a `git fetch` myself with `CI_JOB_TOKEN` or `GITLAB_TOKEN` in the `before_script` it just works:

```bash
$ git fetch "https://gitlab-ci-token:${CI_JOB_TOKEN}@gitlab.com/REDACTED.git"
From https://gitlab.com/REDACTED
 * branch            HEAD       -> FETCH_HEAD
$ git fetch "https://gitlab-ci-token:${GITLAB_TOKEN}@gitlab.com/REDACTED.git"
From https://gitlab.com/REDACTED
 * branch            HEAD       -> FETCH_HEAD
 ```

**Comment by kilianpaquier** [comment]
Thank you for this investigation.

You were saying that for another project it was working fine ? Can you confirm that's the same CI configuration and that the `GITLAB_TOKEN` has the appropriate rights for both projects if shared or the appropriate rights for each projects if not shared ?

You were saying that `git` was installed on your runner, are you using the runner in shell mode or in docker mode ? If in docker mode, did you ensure that git is installed in the docker image either by default or during the CI ?

If not too confidential on your end, would you mind sharing here the release job configured in `.gitlab-ci.yml` ?

**Comment by syphernl** [comment]
> Thank you for this investigation.
> 
> You were saying that for another project it was working fine ? 

Yes, a standalone project (very barebones) works fine. The original project and a fork of the project are experiencing these issues with backmerges. But the pipeline config and branch/tag/variables are configured the same for each of them.

> Can you confirm that's the same CI configuration and that the `GITLAB_TOKEN` has the appropriate rights for both projects if shared or the appropriate rights for each projects if not shared ?

Yes, the token has appropriate scopes. All 3 projects (original, work, barebones test) have their own tokens with the same permissions.

> You were saying that `git` was installed on your runner, are you using the runner in shell mode or in docker mode ? If in docker mode, did you ensure that git is installed in the docker image either by default or during the CI ?

We have a custom image with semantic-release and git included.

> Are you using `CI_JOB_TOKEN` or a custom generated `GITLAB_TOKEN` ?

The plugin and semantic release both require/use the `GITLAB_TOKEN`. Since the 
`CI_JOB_TOKEN` gets masked I cannot verify what is in the `[secret]` portion of the URL.

I have tried to update the git remote prior to running semantic release, to incorporate the `GITLAB_TOKEN` in the URL but it makes no difference.

**Comment by kilianpaquier** [comment]
If not too confidential, could you provide here the semantic-release job configuration in your `.gitlab-ci.yml` please ?

**Comment by kilianpaquier** [comment]
Hi so I tried to reproduce the issue here https://gitlab.com/kilianpaquier/semantic-release-testing without success:
- with a `GITLAB_TOKEN`, masked, hidden and non-protected, with `Developer` role and only `api` scope, backmerge works on a non-protected branch (`staging` to `develop`)
- with a `GITLAB_TOKEN`, masked and hidden, with `Maintainer` role and all `api`, `read_repository` and `write_repository`, backmerge works on a protected branch (`main` to `develop`)

From this quote, it seems `@semantic-release/git` is also impacted like the provided token doesn't have enough rights ...
> In the `before_script` I added the install command, but the error with `marked-terminal` is still the same.
> 
> Could this ENOENT problem be permission related? If it is using the output from `git ls-remote --heads` it will get an URL which Gitlab CI used which includes the (short-lived and limited) [CI_JOB_TOKEN](https://docs.gitlab.co.jp/ee/ci/jobs/ci_job_token.html) that does not have write permissions, while the `GITLAB_TOKEN ` is expected to have that. Still weird, since the error it claims to be failing on is `git fetch` which is allowed via the `CI_JOB_TOKEN`..
> 
> Running with env `DEBUG=semantic-release:*` didn't show any info for the plugin (which was expected as you indicated earlier) but I did see this:
> 
> ```shell
> 2024-10-21T11:46:22.017Z semantic-release:get-git-auth-url Verifying ssh auth by attempting to push to  https://gitlab-ci-token:[secure]@gitlab.com/REDACTED.git
> 2024-10-21T11:46:22.510Z semantic-release:git ExecaError: Command failed with exit code 128: git push --dry-run --no-verify 'https://gitlab-ci-token:[secure]@gitlab.com/REDACTED.git' 'HEAD:acceptance'
> remote: You are not allowed to push code to this project.
> fatal: unable to access 'https://gitlab.com/REDACTED.git/': The requested URL returned error: 403
>     at getFinalError (file:///usr/local/lib/node_modules/semantic-release/node_modules/execa/lib/return/final-error.js:6:9)
>     at makeError (file:///usr/local/lib/node_modules/semantic-release/node_modules/execa/lib/return/result.js:108:16)
>     at getAsyncResult (file:///usr/local/lib/node_modules/semantic-release/node_modules/execa/lib/methods/main-async.js:167:4)
>     at handlePromise (file:///usr/local/lib/node_modules/semantic-release/node_modules/execa/lib/methods/main-async.js:150:17)
>     at async verifyAuth (file:///usr/local/lib/node_modules/semantic-release/lib/git.js:206:5)
>     at async default (file:///usr/local/lib/node_modules/semantic-release/lib/get-git-auth-url.js:93:5)
>     at async run (file:///usr/local/lib/node_modules/semantic-release/index.js:67:27)
>     at async Module.default (file:///usr/local/lib/node_modules/semantic-release/index.js:278:22)
>     at async default (file:///usr/local/lib/node_modules/semantic-release/cli.js:55:5)
> 2024-10-21T11:46:22.511Z semantic-release:get-git-auth-url SSH key auth failed, falling back to https.
> 2024-10-21T11:46:30.600Z semantic-release:get-tags found tags for branch review: []
> 2024-10-21T11:46:31.097Z semantic-release:get-tags found tags for branch acceptance: [ <list_of_tags_here> ]
> 2024-10-21T11:46:31.490Z semantic-release:get-tags found tags for branch production: [ <list_of_tags_here> ]
> [11:46:31 AM] [semantic-release] › ✔  Run automated release from branch acceptance on repository https://gitlab-ci-token:[secure]@gitlab.com/REDACTED.git
> [11:46:32 AM] [semantic-release] › ✔  Allowed to push to the Git repository
> ```

From our discussion, you are talking about `CI_JOB_TOKEN` too, backmerge plugin doesn't read it in its code, in which way are you using this token ?

**Comment by kilianpaquier** [comment]
Thanks ! Could you try to change this line `semantic-release --repository-url="${CI_REPOSITORY_URL}" ${dry_run_opt} | tee semantic-output.txt` to `semantic-release ${dry_run_opt} | tee semantic-output.txt` (removing the repository-url override) ?

**Comment by syphernl** [comment]
I have tested _without_ the `--repository-url` but it makes no difference, it still fails with the same error.  
It does appear I can omit that argument from my template though, because Semantic-Release gets it from `package.json` and/or generates it since the behavior is still the same.

**Comment by kilianpaquier** [comment]
Yes, by default `repository-url` is either retrieved from `package.json` or from git origin URL (which in CI is expected to exist).

If not too confidential too, could you provide your `.releaserc` configuration so I can try reproducing the error on my end with the same conditions ?

**Comment by syphernl** [comment]
Here is our `.releaserc.js`:

```js
module.exports = {
  plugins: [
    [
      '@semantic-release/commit-analyzer',
      {
        preset: 'angular',
        parserOpts: {
          noteKeywords: ['BREAKING CHANGE', 'BREAKING CHANGES', 'BREAKING']
        },
        releaseRules: [
          { type: 'build', release: 'patch' },
          { type: 'deploy', release: 'patch' },
          { type: 'chore', release: 'patch' },
          { type: 'docs', release: 'patch' },
          { type: 'test', release: 'patch' },
          { type: 'style', release: 'patch' },
          { type: 'ci', release: 'patch' }
        ]
      }
    ],
    [
      '@semantic-release/release-notes-generator',
      {
        preset: 'conventionalcommits',
        parserOpts: {
          noteKeywords: ['BREAKING CHANGE', 'BREAKING CHANGES', 'BREAKING']
        },
        writerOpts: {
          commitsSort: ['subject', 'scope']
        },
        presetConfig: {
          types: [
            { type: 'feat', section: 'Features' },
            { type: 'fix', section: 'Bug Fixes' },
            { type: 'chore', section: 'Chores', hidden: false },
            { type: 'refactor', section: 'Internal', hidden: false },
            { type: 'perf', section: 'Performance', hidden: false },
            { type: 'docs', section: 'Documentation', hidden: false },
            { type: 'ci', section: 'DevOps', hidden: false },
            { type: 'test', section: 'Tests', hidden: false }
          ]
        }
      }
    ],
    [
      '@bpgeck/semantic-release-kaniko',{ }
    ],
    [
      '@semantic-release/gitlab',
      {
        successComment: ':tada: This ${issue ? "issue" : "MR"} is included in version [v${nextRelease.version}](${releases.find(release => /gitlab\\.com/i.test(release.url)).url}).',
        labels: 'semantic-release'
      }
    ],
    [
      'semantic-release-slack-bot',
      {
        notifyOnSuccess: true,
        markdownReleaseNotes: true,
        onSuccessTemplate: {
          blocks: [
            {
              type: 'section',
              text: {
                type: 'mrkdwn',
                text: `A new version of \`$package_name\` has been released!\nCurrent version is *$npm_package_version*`
              }
            },
            { type: 'divider' },
            {
              type: 'section',
              text: {
                type: 'mrkdwn',
                text: '$release_notes'
              }
            }
          ],
          attachments: [
            {
              blocks: [
                {
                  type: 'context',
                  elements: [
                    {
                      type: 'mrkdwn',
                      text: `:link: <${process.env.CI_ENVIRONMENT_URL}|View ${process.env.CI_ENVIRONMENT_NAME} website>`
                    }
                  ]
                }
              ]
            }
          ]
        }
      }
    ],
    [
      "@kilianpaquier/semantic-release-backmerge",
      {
        targets: [
          { from: "production", to: "develop" },
          { from: "acceptance", to: "develop" }
        ],
        title: "ci(release): fix mergeback from ${ from } into ${ to }",
      }
    ],
  ],
  branches: [
    { name: 'main', prerelease: true },
    { name: 'review', prerelease: true },
    { name: 'acceptance', prerelease: true },
    { name: 'production', prerelease: false }
  ]
};
```

I have also [reported the issue with marked-terminal](https://github.com/semantic-release/semantic-release/issues/3484) to the Semantic-Release project since that particular problem probably lies there.

This is such a strange issue.. 

* Project repo: fails
* POC repo (I created before applying the setup to our project repo: works fine
* Fork of the project repo: fails
* Newly created "barebones" project: works

It feels like that there is some (UI-)hidden GitLab setting that is active on the project repo and was also added to the fork.. My next step to try is to create a fresh project (no fork) and adding in the project code to see what happens..

**Comment by syphernl** [comment]
> My next step to try is to create a fresh project (no fork) and adding in the project code to see what happens..

Sadly, this causes the same problem..

* Created new a fresh project on GitLab, cloned locally (as `develop`)
* Exported the original repo (as tgz via the GitLab UI), extracted in the new repo folder. Add files, commit (with `[skip ci]`)
* Created Access Token with appropriate access scopes
* Created `acceptance` and `production` branches based on `develop`
* Made them protected (only Maintainer + Access Token can access it)
* Made a "hotfix" on the `acceptance` branch
* Backmerge failed with same ENOENT error

My investigation continues..

**Comment by kilianpaquier** [comment]
Hi !

Could we clarify a little the default values for protected branches accesses for your projects ?

Here's an example for [semantic-release-testing](https://gitlab.com/kilianpaquier/semantic-release-testing):

<img width="617" alt="image" src="https://github.com/user-attachments/assets/d7f97d0a-fe06-4d67-b475-1614e6256eaf">

**Comment by syphernl** [comment]
![image](https://github.com/user-attachments/assets/8d985f0f-a69b-4b4a-9d53-f94aca6f09ea)

The "1 role" is Maintainer and the "1 user" is the Access Token (which is also a Maintainer, so having it explicitly added is a bit redundant)

I am currently testing with some patches to your code to drop in some crude debug logging to figure out what inputs it uses in its functions etc.

**Comment by kilianpaquier** [comment]
It's weird I agree since in the `.gitlab-ci.yml` job, you're executing `git` commands just before ...

But it would explain things, it seems `ENOENT` indicates missing binary in Alpine OS. How is installed **git** with your custom Docker image ?

**Comment by kilianpaquier** [comment]
Do you want me to check it ? You can open a draft pull request 🙂.

Perhaps, if you can, could you try basing the custom image on `node:lts-slim` ? It's 15MB heavier

**Comment by kilianpaquier** [comment]
Do you want me to check it ? You can open a draft pull request 🙂.

Perhaps, if you can, could you try basing the custom image on `node:lts-slim` ? It's 15MB heavier

It's installed with just an `apt-get install git`

**Comment by syphernl** [comment]
> Could you try on your end to use `node:lts-slim` ?

That is something I will test, however it will require quite some extensive rebuilding of the `Dockerfile` since we install a bunch of dependencies which are included in Alpine and not with Debian. 
For actual usage we do prefer to keep using Alpine though. Not just because they are smaller in size, also because the footprint is smaller and results in less potential security issues (have to keep the security scanning jobs happy 😉 )

**Comment by syphernl** [comment]
The Debian-based image is running in to the very same ENOENT issue.. (git version `2.39.5`)

**Comment by kilianpaquier** [comment]
Yes, all git commands should be made with the GITLAB_TOKEN. It could explain the error.

**Comment by syphernl** [comment]
> Yes, all git commands should be made with the GITLAB_TOKEN (repository URL is parsed and authentication is replaced, maybe too late ?). It could explain the error.

The repository URL that gets passed along to `fetch` function contains a `glcbt-xxx` password which is a (temporary) job token and not the `glpat-xxx` token I have configured as `GITLAB_TOKEN`.

**Comment by syphernl** [comment]
I have tested with 1.3.3, the fetch function now does get called with the `glpat-xxx` token so that looks good. 

However... the ENOENT problem still persists.

I have pushed my containers to the GitHub Container Registry. 

-  `ghcr.io/syphernl/semantic-release-test-image:debian`
-  `ghcr.io/syphernl/semantic-release-test-image:alpine`

I am (still) using the `debian` image for testing, the `Dockerfile` for this one:

```Dockerfile
FROM node:20.18.0-slim

# renovate: datasource=github-releases depName=chartmuseum/helm-push
ENV HELMPUSH_VERSION=v0.10.4

# Copy Kaniko executor from the Kaniko image
COPY --chmod=0775 --from=gcr.io/kaniko-project/executor:v1.23.2-slim /kaniko/executor /kaniko/executor

# Install required packages and add Docker's official GPG key and repository
RUN apt-get update && \
    apt-get install -y --no-install-recommends \
        ca-certificates \
        curl \
        jq \
        git \
        gnupg2 && \
    # Add Docker's official GPG key
    curl -fsSL https://download.docker.com/linux/debian/gpg -o /etc/apt/keyrings/docker.asc && \
    # Add Docker repository for bookworm
    echo "deb [arch=amd64 signed-by=/etc/apt/keyrings/docker.asc] https://download.docker.com/linux/debian bookworm stable" > /etc/apt/sources.list.d/docker.list

# Install Docker CLI
RUN apt-get update && \
    apt-get install -y docker-ce-cli && \
    rm -rf /var/lib/apt/lists/*

# Install helm binary
RUN curl -L https://get.helm.sh/helm-v3.16.2-linux-amd64.tar.gz -o helm.tar.gz && \
    tar -zxvf helm.tar.gz && \
    mv linux-amd64/helm /usr/local/bin/helm && \
    chmod +x /usr/local/bin/helm && \
    rm -rf helm.tar.gz linux-amd64

# Install helm-push plugin
RUN helm plugin install https://github.com/chartmuseum/helm-push --version "${HELMPUSH_VERSION}" || true

# Install buildx plugin for Docker
# renovate: datasource=github-releases depName=docker/buildx
ENV BUILDX_VERSION=v0.17.1
RUN curl -L https://github.com/docker/buildx/releases/download/${BUILDX_VERSION}/buildx-${BUILDX_VERSION}.linux-amd64 -o /usr/local/bin/docker-buildx && \
    chmod +x /usr/local/bin/docker-buildx

WORKDIR /

# Copy over dependencies needed by npm
COPY package.json package-lock.json /

# Install semantic-release plugins listed in package.json
RUN jq -r '.dependencies | to_entries | map("\(.key)@\(.value)") | .[]' /package.json > /deps.txt && \
    npm install -g --unsafe-perm $(cat /deps.txt) && \
    rm -f /deps.txt /package.json /package-lock.json && \
    npm cache clean --force

WORKDIR /workspace

ENTRYPOINT [ "/usr/local/bin/semantic-release" ]
CMD ["--help"]
```

Dependencies in `package.json`:

```json
    "dependencies": {
        "@bpgeck/semantic-release-kaniko": "2.0.2",
        "@kilianpaquier/semantic-release-backmerge": "1.3.3",
        "@semantic-release/exec": "6.0.3",
        "@semantic-release/gitlab": "13.2.1",
        "conventional-changelog-conventionalcommits": "8.0.0",
        "semantic-release": "<25",
        "semantic-release-helm3": "2.9.3",
        "semantic-release-slack-bot": "4.0.2",
        "@codedependant/semantic-release-docker": "5.0.3"
    }
```

---

I replace the `index.js` with one on Github gists with additional debugging in the `before_script` and as a test, I added this before the `fetch()`:

```ts
    // Show the PATH
    context.logger.log(`DEBUG: System PATH: ${process.env.PATH}`);

    // Check if git exists
    exec("git --version", (error: Error | null, stdout: string, stderr: string) => {
    if (error) {
        context.logger.error("DEBUG: Git is not installed or not in PATH.");
        return;
    }

    if (stderr) {
        context.logger.error(`DEBUG: Error checking Git version: ${stderr}`);
        return;
    }

    context.logger.log(`DEBUG: Git is installed. Version: ${stdout.trim()}`);
    });
```

It reports ```DEBUG: Git is not installed or not in PATH```

**Comment by syphernl** [comment]
Some further investigations... I have patched the `execaSync` function calls to include `shell: true` in the options and it results in a slightly different error message:

```bash
Failed to backmerge branches.
xJ: Command failed with exit code 127: git fetch 'https://gitlab-ci-token:[secure]@gitlab.com/[REDACTED].git'
/bin/sh: git: not found
```

- My own `git` commands in `scripts:` or `before_scripts` work fine
- `Semantic-release/git` also [uses](https://github.com/semantic-release/git/blob/master/lib/git.js) the `git` binary directly and `execa`, which calls commands before reaching the backmerge plugin. 
- Same for [semantic-release itself](https://github.com/semantic-release/semantic-release/blob/master/lib/git.js)

The first difference that  I can see is that `semantic-release` / `semantic-release/git` both use `execa` while this plugin uses `execaSync`

As a test I replaced the `execaSync` with `await execa` which results in another different (but similar) error message:

```bash
NJ: Command failed with ENOENT: git fetch 'https://gitlab-ci-token:[MASKED]@gitlab.com/[REDACTED].git'
spawn git ENOENT
    at Tq (file:///usr/local/lib/node_modules/@kilianpaquier/semantic-release-backmerge/dist/index.js:12:12934)
    at w2 (file:///usr/local/lib/node_modules/@kilianpaquier/semantic-release-backmerge/dist/index.js:25:12494)
    at Ax (file:///usr/local/lib/node_modules/@kilianpaquier/semantic-release-backmerge/dist/index.js:28:38248)
    at xx (file:///usr/local/lib/node_modules/@kilianpaquier/semantic-release-backmerge/dist/index.js:28:37967)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async eX (file:///usr/local/lib/node_modules/@kilianpaquier/semantic-release-backmerge/dist/index.js:28:41002) {
  shortMessage: "Command failed with ENOENT: git fetch 'https://gitlab-ci-token:[MASKED]@gitlab.com/[REDACTED].git'\n" +
    'spawn git ENOENT',
  originalMessage: 'spawn git ENOENT',
  command: 'git fetch https://gitlab-ci-token:[MASKED]@gitlab.com/[REDACTED].git',
  escapedCommand: "git fetch 'https://gitlab-ci-token:[MASKED]@gitlab.com/[REDACTED].git'",
  cwd: '/builds/[REDACTED]',
  durationMs: 25.036707,
  failed: true,
  timedOut: false,
  isCanceled: false,
  isGracefullyCanceled: false,
  isTerminated: false,
  isMaxBuffer: false,
  isForcefullyTerminated: false,
  code: 'ENOENT',
  stdout: '',
  stderr: '',
  stdio: [ undefined, '', '' ],
  ipcOutput: [],
  pipedFrom: [],
  [cause]: Error: spawn git ENOENT
      at ChildProcess._handle.onexit (node:internal/child_process:286:19)
      at onErrorNT (node:internal/child_process:484:16)
      at process.processTicksAndRejections (node:internal/process/task_queues:82:21) {
    errno: -2,
    code: 'ENOENT',
    syscall: 'spawn git',
    path: 'git',
    spawnargs: [
      'fetch',
      'https://gitlab-ci-token:[MASKED]@gitlab.com/[REDACTED].git'
    ]
  }
}
```

**Comment by kilianpaquier** [comment]
Hi !

It could come from input `env`, did you try printing it in your investigation ?

Maybe this `env` variable contains / not contains the right `$PATH` with your installed git path.

**Comment by syphernl** [comment]
The `env` output is massive, it contains vars from the container and Gitlab injected ones. The path is correct though, but the command is still failing even if the path is explicitly set to `/usr/bin/git`

**Comment by syphernl** [comment]
I wanted to make sure our custom runners were not the problem, so I also tested the pipeline against the GitLab.com SaaS runners but this makes no difference.

> Are you using `await execa` with or without `shell` option right now for your tests ? Did you try with and without this option provided to see if anything would change ?

Initially I tested without the `shell` option. With `{shell: true}` this results in:

```bash
NJ: Command failed with exit code 127: git ls-remote --heads 'https://gitlab-ci-token:[MASKED]@gitlab.com/[REDACTED].git'
/bin/sh: git: not found
    at Tq (file:///usr/local/lib/node_modules/@kilianpaquier/semantic-release-backmerge/dist/index.js:12:12934)
    at w2 (file:///usr/local/lib/node_modules/@kilianpaquier/semantic-release-backmerge/dist/index.js:25:12494)
    at Ax (file:///usr/local/lib/node_modules/@kilianpaquier/semantic-release-backmerge/dist/index.js:28:38248)
    at xx (file:///usr/local/lib/node_modules/@kilianpaquier/semantic-release-backmerge/dist/index.js:28:37967)
    at async _B (file:///usr/local/lib/node_modules/@kilianpaquier/semantic-release-backmerge/dist/index.js:28:40710) {
  shortMessage: "Command failed with exit code 127: git ls-remote --heads 'https://gitlab-ci-token:[MASKED]@gitlab.com/[REDACTED].git'",
  command: 'git ls-remote --heads https://gitlab-ci-token:[MASKED]@gitlab.com/[REDACTED].git',
  escapedCommand: "git ls-remote --heads 'https://gitlab-ci-token:[MASKED]@gitlab.com/[REDACTED].git'",
  cwd: '/builds/[REDACTED]',
  durationMs: 16.409128,
  failed: true,
  timedOut: false,
  isCanceled: false,
  isGracefullyCanceled: false,
  isTerminated: false,
  isMaxBuffer: false,
  isForcefullyTerminated: false,
  exitCode: 127,
  stdout: '',
  stderr: '/bin/sh: git: not found',
  stdio: [ undefined, '', '/bin/sh: git: not found' ],
  ipcOutput: [],
  pipedFrom: []
}
Node.js v20.18.0
```

**Comment by kilianpaquier** [comment]
It seems to have failed on branch listing and not fetch (after fetch as such), did both git commands had the same configuration ?

**Comment by syphernl** [comment]
> It seems to have failed on branch listing and not fetch (after fetch as such), did both git commands had the same configuration ?

I noticed that it was `ls` and not `fetch` that failed this time.. But that means that `ls` would work normally and just the `fetch` would fail? 🤔 

This is the code I used for the previous run:

```ts
import { execa } from "execa"

/**
 * ls returns the slice of all branches present in remote origin.
 * It removes 'refs/heads/' from the branches name.
 *
 * @returns the slice of branches.
 *
 * @throws an error if the git ls-remote cannot be done.
 */
export const ls = async (remote: string, cwd?: string, env?: Record<string, string>) => {
    const { stdout } = await execa("git", ["ls-remote", "--heads", remote], {shell: true})
    const branches = stdout.
        split("\n").
        map(branch => branch.split("\t")).
        flat().
        filter(branch => branch.startsWith("refs/heads/")).
        map(branch => branch.replace("refs/heads/", ""))
    return [...new Set(branches)]
}

/**
 * checkout executes a simple checkout of input branch.
 * The checkout is strict with remote state, meaning all local changes are removed.
 *
 * @param branch the input branch to checkout.
 *
 * @throws an error if the checkout cannot be done.
 */
export const checkout = async (branch: string, cwd?: string, env?: Record<string, string>) => {
    await execa("git", ["checkout", "-B", branch], {shell: true})
}

/**
 * fetch executes a simple fetch of input remote.
 *
 * @param remote the remote to fetch branches from.
 *
 * @throws an error if the fetch cannot be done.
 */
export const fetch = async (remote: string, cwd?: string, env?: Record<string, string>) => {
    await execa("git", ["fetch", remote], {shell: true})
}

/**
 * merge executes a checkout of input 'to' branch, and merges input 'from' branch into 'to'.
 * If a merge commit must be done (by default --ff is used), then the merge commit is the input commit.
 *
 * @param from the branch to merge into 'to'.
 * @param to the branch to merge changes from 'from'.
 * @param commit the merge commit message if one is done.
 *
 * @throws an error if the merge fails (in case of conflicts, etc.).
 */
export const merge = async (from: string, to: string, commit: string, cwd?: string, env?: Record<string, string>) => {
    await checkout(to)

    try {
        await execa("git", ["merge", `${from}`, "--ff", "-m", commit], {shell: true})
    } catch (error) {
        await execa("git", ["merge", "--abort"], {shell: true})
        throw error
    }
}

/**
 * push executes a simple git push to the input remote with the current checked out branch.
 *
 * @param remote the remote to push changes to.
 * @param dryRun if the push must only verify if all conditions are fine and not alter the remote state.
 *
 * @throws an error if the push cannot be executed.
 */
export const push = async (remote: string, branch: string, dryRun?: boolean, cwd?: string, env?: Record<string, string>) => {
    const args = ["push", remote, `HEAD:${branch}`]
    if (dryRun) {
        args.push("--dry-run")
    }
    await execa("git", args, {shell: true})
}
```

**Comment by kilianpaquier** [comment]
`fetch` is executed before `ls`, if it failed on `ls` it means that `fetch` was fine.

But with the provided code, I don't see any difference that would explain `fetch` working.

**Comment by kilianpaquier** [comment]
You could try without shell, but I'm not sure, maybe with `authRemote`, ls fails with an stderr text but still returns something in stdout without couldn't be parsed as a list ?

Did you try using backmerge version `v1.1.5` ? It's using an older version of execa (v8) whereas newest versions use execa v9 (in cas it's a spawn error). Features added between v1.1.5 and v1.3.3 should block (only fetch is made with `repositoryUrl` instead of `authRemote`)

**Comment by syphernl** [comment]
**I have some good news! 🎉** 

It looks like replacing `execaSync` with `await execa` is making things work. The process works on both `alpine` and `debian` images, without needing `shell: true`. 

However.... the backmerge itself seems to fail. 

As test:

* I made a simple change to the `acceptance` branch, expecting it to backmerge that to `develop`
* It fails with `Updates were rejected because the tip of your current branch is behind its remote counterpart`
* It opened a MR instead, with the changes done to `acceptance` and no merge conflicts or anything

But the good news: executing the `git` commands works fine when using `execa` instead of `execaSync`! 
This is the same way of interacting with `git`  as `semantic-release/semantic-release` and `semantic-release/git` is doing., so to me it would make sense to switch to this method.

I would open a PR to deal with this but my code has become _extremely_ messy, riddled with debug code and I probably over-complicated some things as well (as I am not a JS dev nor have much async experience)

**Comment by kilianpaquier** [comment]
I can fix the code and add you as co-developer if you need to reorder things on your end after this big investigation :)

Yeah this error means that git wasn't able to merge your branch, either because there're conflicts or because your target backmerged branch as newer commits. If it opens the pull request it's fine ^^

**Comment by syphernl** [comment]
> Did you try using backmerge version `v1.1.5` ? It's using an older version of execa (v8) whereas newest versions use execa v9 (in cas it's a spawn error). Features added between v1.1.5 and v1.3.3 should block (only fetch is made with `repositoryUrl` instead of `authRemote`).
> 
> `semantic-release/git` is using execa v5 and semantic-release itself is using `v9.0.0`

I just tested with `v1.1.5`  (without _any_ patches/changes) on my Alpine image and it behaves the same as my patched code. 
I am also getting the `Updates were rejected because the tip of your current branch is behind` error with that version.

> Yeah this error means that git wasn't able to merge your branch, either because there're conflicts or because your target backmerged branch as newer commits. If it opens the pull request it's fine ^^

Is it really fine? Because with both my patched version and `v1.1.5` are failing to backmerge a simple change and reverts to a PR instead (which merges just fine). 
I really hope we haven't stumbled upon a new problem here 😅 

> Could you still maybe open a draft pull request so I can check what changes made the whole thing work correctly ?

I'll create a draft PR based on my changes.

**Comment by kilianpaquier** [comment]
I'll check specifically this second issue, it's easier to reproduce on my testing project and already encountered it ahah.

Ok so it really was execaSync the issue ... Since v1.1.5 was use `await execa` (if I remember well)

**Comment by syphernl** [comment]
I created the draft PR: https://github.com/kilianpaquier/semantic-release-backmerge/pull/24

> I'll check specifically this second issue, it's easier to reproduce on my testing project and already encountered it ahah.

Would definitely help if you can reproduce it more easily than this particular issue 😅 

> Ok so it really was execaSync the issue ... Since v1.1.5 was use `await execa` (if I remember well)

Yep looks like it. Out of curiosity, why did you decide to switch over to `execAsync`? 

I it still not clear as to why it would only be problematic on certain projects and not all, using the very same config/containers etc. For 2 test (reproduction) projects it worked just fine, for the actual project it fails. 
Perhaps that will remain a mystery forever... 

At least the issue has been uncovered and there is a fix for it on the horizon so I can make our devs happy with a pipeline that will complete successfully instead of causing panic that "the release failed" while it was only the backmerge that did 😉

**Comment by kilianpaquier** [comment]
Ahah just because it avoided a lot of `async` before functions. JS is neither on my end my predilection langage and I was just "yeah, less async, preferred on my end"

>Yep looks like it. Out of curiosity, why did you decide to switch over to execAsync?

**Comment by kilianpaquier** [comment]
Could you provide me with your github email (`Co-authored-by: syphernl <ID+syphernl@users.noreply.github.com>`) for co-authoring ? :)

See https://docs.github.com/fr/account-and-profile/setting-up-and-managing-your-personal-account-on-github/managing-email-preferences/setting-your-commit-email-address
